File: app/src/main/java/com/mapmyindia/sdk/demo/java/activity/SafetyPluginActivity.java
package com.mapmyindia.sdk.demo.java.activity;

import android.os.Bundle;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.mapmyindia.sdk.demo.R;
import com.mapmyindia.sdk.safety.MapmyIndiaSafetyPlugin;
import com.mapmyindia.sdk.safety.callbacks.ISafetyListener;
import com.mapmyindia.sdk.safety.containmentlayer.vo.ContainmentZoneInfo;
import com.mapmyindia.sdk.safety.internal.callbacks.IAuthListener;

import java.text.DecimalFormat;

import timber.log.Timber;

public class SafetyPluginActivity extends AppCompatActivity implements View.OnClickListener {


    private TextView txtInsideZone;
    private TextView txtZoneName;
    private TextView txtDistance;
    private TextView txtMapLink;
    private TextView txtDistrictName;
    private TextView txtZoneType;
    private TextView startStopSafetyPlugin;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_safety_plugin);


        findViewById(R.id.initialize_btn).setOnClickListener(this);
        findViewById(R.id.get_containment_zone_btn).setOnClickListener(this);
        startStopSafetyPlugin = findViewById(R.id.start_safety_plugin_btn);
        startStopSafetyPlugin.setOnClickListener(this);
        txtInsideZone = findViewById(R.id.txt_inside_zone);
        txtZoneName = findViewById(R.id.txt_zone_name);
        txtDistance = findViewById(R.id.txt_distance);
        txtMapLink = findViewById(R.id.txt_map_link);
        txtDistrictName = findViewById(R.id.txt_district_name);
        txtZoneType = findViewById(R.id.txt_zone_type);
        findViewById(R.id.initialize_btn).setEnabled(!MapmyIndiaSafetyPlugin.getInstance().isInitialized());
        startStopSafetyPlugin.setText(!MapmyIndiaSafetyPlugin.getInstance().isRunning() ? R.string.txt_start_safety_plugin : R.string.txt_stop_safety_plugin);


    }

    private void reset() {
        txtInsideZone.setText("Inside Containment Zone: ");
        txtDistance.setText("Nearest zone distance: ");
        txtMapLink.setText("Containment Zone Link: ");
        txtDistrictName.setText("District Name: ");
        txtZoneType.setText("Zone Type: ");
    }

    private String getDistanceFormat(long distance) {
        if (distance >= 1000) {
            double dist = (distance / 1000.0);
            String distFinal = new DecimalFormat("#.0").format(dist);
            return distFinal + "Km(s)";
        } else {
            int dist = (int) (distance);
            return this.getResources().getQuantityString(R.plurals.distance_meter, dist, dist);
        }
    }


    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.initialize_btn:

                MapmyIndiaSafetyPlugin.getInstance().initialize(new IAuthListener() {
                    @Override
                    public void onSuccess() {

                        Toast.makeText(SafetyPluginActivity.this, "onSuccess", Toast.LENGTH_SHORT).show();
                        view.setEnabled(!MapmyIndiaSafetyPlugin.getInstance().isInitialized());
                    }

                    @Override
                    public void onError(String reason, String errorIdentifier, String errorDescription) {
                        Toast.makeText(SafetyPluginActivity.this, "onFail:- " + reason, Toast.LENGTH_SHORT).show();
                    }
                });
                break;
            case R.id.get_containment_zone_btn:
                reset();
                MapmyIndiaSafetyPlugin.getInstance().getCurrentLocationSafety(new ISafetyListener() {
                    @Override
                    public void onResult(ContainmentZoneInfo zoneInfo) {
                        txtInsideZone.setText(String.format("Inside Containment Zone: %s", zoneInfo.isInsideContainmentZone()));
                        txtDistance.setText(String.format("Nearest zone distance: %s", getDistanceFormat(zoneInfo.getDistanceToNearestZone())));
                        txtMapLink.setText(String.format("Containment Zone Link: %s", zoneInfo.getMapLink()));
                        txtDistrictName.setText(String.format("District Name: %s", zoneInfo.getDistrictName()));
                        txtZoneType.setText(String.format("Zone Type: %s", zoneInfo.getZoneType()));

                    }

                    @Override
                    public void onError(String reason, String errorIdentifier, String errorDescription) {
                        Timber.e("OnError:" + reason + "errorIdentifier:- " + errorIdentifier);
                        Toast.makeText(SafetyPluginActivity.this, "" + errorDescription, Toast.LENGTH_SHORT).show();
                    }
                });
                break;
            case R.id.start_safety_plugin_btn:

               
                reset();
                if (!MapmyIndiaSafetyPlugin.getInstance().isRunning()) {
                    MapmyIndiaSafetyPlugin.getInstance().startSafetyPlugin(new ISafetyListener() {
                        @Override
                        public void onResult(ContainmentZoneInfo zoneInfo) {
                            startStopSafetyPlugin.setText(!MapmyIndiaSafetyPlugin.getInstance().isRunning() ? R.string.txt_start_safety_plugin : R.string.txt_stop_safety_plugin);
                            txtInsideZone.setText(String.format("Inside Containment Zone: %s", zoneInfo.isInsideContainmentZone()));
                            txtDistance.setText(String.format("Nearest zone distance: %s", getDistanceFormat(zoneInfo.getDistanceToNearestZone())));
                            txtMapLink.setText(String.format("Containment Zone Link: %s", zoneInfo.getMapLink()));
                            txtDistrictName.setText(String.format("District Name: %s", zoneInfo.getDistrictName()));
                            txtZoneType.setText(String.format("Zone Type: %s", zoneInfo.getZoneType()));
                        }

                        @Override
                        public void onError(String reason, String errorIdentifier, String errorDescription) {
                            Timber.e("OnError:" + reason + "errorIdentifier:- " + errorIdentifier);
                        }
                    });
                } else {
                    MapmyIndiaSafetyPlugin.getInstance().stopSafetyPlugin();
                    startStopSafetyPlugin.setText(!MapmyIndiaSafetyPlugin.getInstance().isRunning() ? R.string.txt_start_safety_plugin : R.string.txt_stop_safety_plugin);

                }
                break;
        }
    }


}

File: app/src/main/java/com/mapmyindia/sdk/demo/kotlin/activity/SafetyPluginActivity.kt
package com.mapmyindia.sdk.demo.kotlin.activity

import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.mapmyindia.sdk.demo.R
import com.mapmyindia.sdk.safety.MapmyIndiaSafetyPlugin
import com.mapmyindia.sdk.safety.callbacks.ISafetyListener
import com.mapmyindia.sdk.safety.containmentlayer.vo.ContainmentZoneInfo
import com.mapmyindia.sdk.safety.internal.callbacks.IAuthListener
import timber.log.Timber
import java.text.DecimalFormat

class SafetyPluginActivity : AppCompatActivity(), View.OnClickListener {
    private var txtInsideZone: TextView? = null
    private var txtZoneName: TextView? = null
    private var txtDistance: TextView? = null
    private var txtMapLink: TextView? = null
    private var txtDistrictName: TextView? = null
    private var txtZoneType: TextView? = null
    private var startStopSafetyPlugin: Button? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_safety_plugin)
        findViewById<View>(R.id.initialize_btn).setOnClickListener(this)
        findViewById<View>(R.id.get_containment_zone_btn).setOnClickListener(this)
        startStopSafetyPlugin = findViewById<Button>(R.id.start_safety_plugin_btn)
        startStopSafetyPlugin?.setOnClickListener(this)
        txtInsideZone = findViewById(R.id.txt_inside_zone)
        txtZoneName = findViewById(R.id.txt_zone_name)
        txtDistance = findViewById(R.id.txt_distance)
        txtMapLink = findViewById(R.id.txt_map_link)
        txtDistrictName = findViewById(R.id.txt_district_name)
        txtZoneType = findViewById(R.id.txt_zone_type)
        findViewById<View>(R.id.initialize_btn).isEnabled = !MapmyIndiaSafetyPlugin.getInstance().isInitialized
        startStopSafetyPlugin?.setText(if (!MapmyIndiaSafetyPlugin.getInstance().isRunning) R.string.txt_start_safety_plugin else R.string.txt_stop_safety_plugin)
    }

    private fun reset() {
        txtInsideZone?.text = "Inside Containment Zone: "
        txtDistance?.text = "Nearest zone distance: "
        txtMapLink?.text = "Containment Zone Link: "
        txtDistrictName?.text = "District Name: "
        txtZoneType?.text = "Zone Type: "
    }

    private fun getDistanceFormat(distance: Long): String {
        return if (distance >= 1000) {
            val dist = distance / 1000.0
            val distFinal = DecimalFormat("#.0").format(dist)
            distFinal + "Km(s)"
        } else {
            val dist = distance.toInt()
            this.resources.getQuantityString(R.plurals.distance_meter, dist, dist)
        }
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.initialize_btn -> MapmyIndiaSafetyPlugin.getInstance().initialize(object : IAuthListener {
                override fun onSuccess() {
                    Toast.makeText(this@SafetyPluginActivity, "onSuccess", Toast.LENGTH_SHORT).show()
                    view.isEnabled = !MapmyIndiaSafetyPlugin.getInstance().isInitialized
                }

                override fun onError(reason: String, errorIdentifier: String, errorDescription: String) {
                    Toast.makeText(this@SafetyPluginActivity, "onFail:- $reason", Toast.LENGTH_SHORT).show()
                }
            })
            R.id.get_containment_zone_btn -> {
                reset()
                MapmyIndiaSafetyPlugin.getInstance().getCurrentLocationSafety(object : ISafetyListener {
                    override fun onResult(zoneInfo: ContainmentZoneInfo) {
                        txtInsideZone!!.text = String.format("Inside Containment Zone: %s", zoneInfo.isInsideContainmentZone)
                        txtDistance!!.text = String.format("Nearest zone distance: %s", getDistanceFormat(zoneInfo.distanceToNearestZone))
                        txtMapLink!!.text = String.format("Containment Zone Link: %s", zoneInfo.mapLink)
                        txtDistrictName!!.text = String.format("District Name: %s", zoneInfo.districtName)
                        txtZoneType!!.text = String.format("Zone Type: %s", zoneInfo.zoneType)
                    }

                    override fun onError(reason: String, errorIdentifier: String, errorDescription: String) {
                        Timber.e("OnError:" + reason + "errorIdentifier:- " + errorIdentifier)
                        Toast.makeText(this@SafetyPluginActivity, "" + errorDescription, Toast.LENGTH_SHORT).show()
                    }
                })
            }
            R.id.start_safety_plugin_btn -> {
                reset()
                if (!MapmyIndiaSafetyPlugin.getInstance().isRunning) {
                    MapmyIndiaSafetyPlugin.getInstance().startSafetyPlugin(object : ISafetyListener {
                        override fun onResult(zoneInfo: ContainmentZoneInfo) {
                            startStopSafetyPlugin!!.setText(if (!MapmyIndiaSafetyPlugin.getInstance().isRunning) R.string.txt_start_safety_plugin else R.string.txt_stop_safety_plugin)
                            txtInsideZone!!.text = String.format("Inside Containment Zone: %s", zoneInfo.isInsideContainmentZone)
                            txtDistance!!.text = String.format("Nearest zone distance: %s", getDistanceFormat(zoneInfo.distanceToNearestZone))
                            txtMapLink!!.text = String.format("Containment Zone Link: %s", zoneInfo.mapLink)
                            txtDistrictName!!.text = String.format("District Name: %s", zoneInfo.districtName)
                            txtZoneType!!.text = String.format("Zone Type: %s", zoneInfo.zoneType)
                        }

                        override fun onError(reason: String, errorIdentifier: String, errorDescription: String) {
                            Timber.e("OnError:" + reason + "errorIdentifier:- " + errorIdentifier)
                        }
                    })
                } else {
                    MapmyIndiaSafetyPlugin.getInstance().stopSafetyPlugin()
                    startStopSafetyPlugin!!.setText(if (!MapmyIndiaSafetyPlugin.getInstance().isRunning) R.string.txt_start_safety_plugin else R.string.txt_stop_safety_plugin)
                }
            }
        }
    }
}